Addition of LLVM.

dada2048 opened the pull request:
This is a proposed addition of the LLVM Compiler Infrastructure (version 4.0.0).
It compiles and works for at least one example program but it needs more testing and maybe some changes (this is my first port, I'm not sure I did everything correctly).

Port request is in issue #508.

dada2048 committed:
Added LLVM.

dada2048 committed:
Reduce size of the build and fix post-build errors.

dada2048 committed:
Merge branch 'master' into add-llvm

dada2048 committed:
Disable ABI breaking checks.

dada2048 commented on ports/llvm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/bin)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/bin)
+
+# Remove one empty include subdirectory
```
Is there a safer way to do this? Any upstream change that populates this directory will require a change to the port...

dada2048 commented on ports/llvm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+
+vcpkg_install_cmake()
+
+# Move cmake modules to correct vcpkg location and remove extra copy
```
I'm not sure those cmake files can handle being installed in a different location, needs testing.

dada2048 commented on ports/llvm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+
+# Remove extra copies of include files in debug directory
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+
```
Is there a safer way to do this? Any upstream change that populates this directory will require a change to the port...

dada2048 commented on ports/llvm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+# Move cmake modules to correct vcpkg location and remove extra copy
+file(COPY ${CURRENT_PACKAGES_DIR}/lib/cmake/llvm DESTINATION ${CURRENT_PACKAGES_DIR}/share)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/lib/cmake)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/lib/cmake)
```
I'm not sure those cmake files can handle being installed in a different location, needs testing.

ras0219-msft replied:
We do have a helper script that performs a bunch of automatic munging as long as the files are pretty standard -- `vcpkg_fixup_cmake_targets()`.

In this case, I think these three lines can be replaced with
```cmake
vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/llvm)
```

dada2048 commented on ports/llvm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/bin)
+
+# Remove one empty include subdirectory
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/include/llvm/MC/MCAnalysis)
```
Is there a safer way to do this? Any upstream change that populates this directory will require a change to the port...

ras0219-msft replied:
We could check if there are contents in the folder before deleting it:
```cmake
file(GLOB X ${CURRENT_PACKAGES_DIR}/include/llvm/MC/MCAnalysis/*)
if(NOT X)
  file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/include/llvm/MC/MCAnalysis)
endif()
```

ras0219-msft commented on ports/llvm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        -DLLVM_TARGETS_TO_BUILD=X86
```
Does this include the entire x86 family (+amd64)? Or is this just i686?

We might want to think about ARM here as well, but that's a longer term issue -- it shouldn't block this initial port.

dada2048 replied:
This is just 32bit X86. I'm open to make it "X86;AArch64" or "X86;AArch64;ARM" although I'm having trouble with the syntax for passing quotes.

But I was actually planning to leave out this option altogether, which builds all architectures (that's the default on all platforms). The reason I have X86 now is that it takes a horrendous time to compile and I wanted to iterate a bit faster.

I will implement your other suggestions this weekend and do some more testing (already have a couple examples running with JIT and everything so I believe I'm on the right track).

dada2048 committed:
Use vcpkg_fixup_cmake_targets to move cmake modules in place.
Make sure MCAnalysis directory is empty before removing.
Use file(INSTALL ...) to put licence file in place.

dada2048 committed:
Merge branch 'master' into add-llvm

dada2048 commented:
I think this is now ready to merge.

Feel free to remove the LLVM_TARGETS_TO_BUILD option if you prefer to have all architectures built by default but the buildtree is already twice the size of boost with just x86 and x64...

dada2048 committed:
Patch CMakeLists to install CMake modules to share/llvm instead of lib/cmake/llvm. This works better than vcpkg_fixup_cmake_targets because other parts of the LLVM build system really want all the variables set correctly. Also set other options to prevent creation of any bin/* instead of removing them after the fact, for the same reason.

dada2048 commented:
Anything else I can do here?

ras0219-msft commented:
Nope, thanks for the PR!

(Sorry for the delay! It's summer vacation season in the US 🌴)

dada2048 commented:
No worries! Thanks for the feedback and for merging.

AraHaan commented:
Was this updated for clang 5.0.0 too? I need it to work with 5.0.0.
